File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/command/currency/dataprovider/nbrb/NBRBExRatesParser.java
package com.github.bumblebee.command.currency.dataprovider.nbrb;

import org.springframework.stereotype.Component;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import java.io.IOException;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

@Component
public class NBRBExRatesParser {

    private static final String API_URL = "http://nbrb.by/Services/XmlExRates.aspx?ondate=";
    private static final String NUM_CODE = "NumCode";
    private static final String NAME = "Name";
    private static final String CHAR_CODE = "CharCode";
    private static final String SCALE = "Scale";
    private static final String RATE = "Rate";
    private static final String CURRENCY = "Currency";

    private DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("MM/dd/yyyy");
    private DocumentBuilder documentBuilder;

    public NBRBExRatesParser() {
        try {
            documentBuilder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
        } catch (ParserConfigurationException e) {
            throw new IllegalStateException(e);
        }
    }

    public List<Currency> getDailyRates() throws IOException, SAXException {
        return getDailyRates(null);
    }

    public List<Currency> getDailyRates(LocalDate date) throws IOException, SAXException {

        Document doc = documentBuilder.parse(getRequestUrl(date));

        List<Currency> currencies = new ArrayList<>();
        NodeList list = doc.getElementsByTagName(CURRENCY);
        for (int i = 0; i < list.getLength(); i++) {
            Node node = list.item(i);
            if (node.getNodeType() == Node.ELEMENT_NODE) {
                currencies.add(convert((Element) node));
            }
        }
        return currencies;
    }

    private String getRequestUrl(LocalDate date) {
        return (date != null) ? API_URL + date.format(dateFormatter) : API_URL;
    }

    private Currency convert(Element currencyElement) {
        Currency currency = new Currency();
        currency.setCode(getValue(NUM_CODE, currencyElement));
        currency.setName(getValue(NAME, currencyElement));
        currency.setShortName(getValue(CHAR_CODE, currencyElement));
        currency.setScale(Integer.parseInt(getValue(SCALE, currencyElement)));
        currency.setRate(Double.parseDouble(getValue(RATE, currencyElement)));
        return currency;
    }

    private String getValue(String tagName, Element e) {
        String value = null;

        NodeList list = e.getElementsByTagName(tagName);
        if (list != null && list.getLength() > 0) {
            NodeList childList = list.item(0).getChildNodes();
            if (childList != null && childList.getLength() > 0) {
                value = childList.item(0).getNodeValue();
            }
        }

        return value;
    }

}

File: telegram-bot-bumblebee-core/src/main/kotlin/com/github/bumblebee/command/currency/dataprovider/nbrb/NBRBExRatesParser.kt
package com.github.bumblebee.command.currency.dataprovider.nbrb

import org.springframework.stereotype.Component
import org.w3c.dom.Element
import org.w3c.dom.Node
import org.xml.sax.SAXException
import java.io.IOException
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import javax.xml.parsers.DocumentBuilderFactory

@Component
class NBRBExRatesParser {

    private val dateFormatter = DateTimeFormatter.ofPattern("MM/dd/yyyy")
    private val documentBuilder = DocumentBuilderFactory.newInstance().newDocumentBuilder()

    fun getDailyRates(): List<Currency> = getDailyRates(null)

    @Throws(IOException::class, SAXException::class)
    fun getDailyRates(date: LocalDate?): List<Currency> {
        val doc = documentBuilder.parse(getRequestUrl(date))

        val list = doc.getElementsByTagName(CURRENCY)
        return (0 until list.length).asSequence()
                .map { list.item(it) }
                .filter { it.nodeType == Node.ELEMENT_NODE }
                .map { convert(it as Element) }
                .toList()
    }

    private fun getRequestUrl(date: LocalDate?): String {
        return if (date != null) API_URL + date.format(dateFormatter) else API_URL
    }

    private fun convert(currencyElement: Element): Currency {
        return Currency(
                code = getValue(NUM_CODE, currencyElement),
                name = getValue(NAME, currencyElement),
                shortName = getValue(CHAR_CODE, currencyElement),
                scale = getValue(SCALE, currencyElement).toInt(),
                rate = getValue(RATE, currencyElement).toDouble()
        )
    }

    private fun getValue(tagName: String, e: Element): String {
        val list = e.getElementsByTagName(tagName)
        if (list != null && list.length > 0) {
            val childList = list.item(0).childNodes
            if (childList != null && childList.length > 0) {
                return childList.item(0).nodeValue
            }
        }
        throw IllegalStateException("Tag not found: $tagName")
    }

    companion object {
        private val API_URL = "http://nbrb.by/Services/XmlExRates.aspx?ondate="
        private val NUM_CODE = "NumCode"
        private val NAME = "Name"
        private val CHAR_CODE = "CharCode"
        private val SCALE = "Scale"
        private val RATE = "Rate"
        private val CURRENCY = "Currency"
    }
}
